oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,492 @@
+
+Approach
+========
+
+We intend to enable runtime module loading and shared dependencies via `Webpack module federation <https://webpack.js.org/concepts/module-federation>`_. More information on module federation beyond its webpack implementation can be found on its `dedicated site at module-federation.io <https://module-federation.io/>`_.
+
+Further, we believe that we also need an approach to maintaining and enforcing dependency consistency across MFEs to realize the benefits of shared dependencies. MFEs need a compatible version of the shared dependency to be available, otherwise they need to fall back to their own version. The process, tooling, and/or code organization necessary to provide that consistency is not something module federation can help with and needs to be addressed separately.
+
+Module Federation
+-----------------
+
+Module federation is implemented as a `plugin for Webpack <https://webpack.js.org/plugins/module-federation-plugin/>`_ that enables micro-frontends to be composed into the same page at runtime even if they're built separately and independently deployed. The pieces being composed are "modules". It lets us configure which dependencies should be shared between modules on a page and what modules a particular frontend exposes to be loaded by other frontends.
+
+If two modules require incompatible versions of a shared dependency, the second one loaded can fall back to loading a version it's compatible with from its own build. (see the link in step #4 below for details.)
+
+In terms of Open edX MFEs, this means:
+
+1. MFEs can continue to be built independently.
+2. The Webpack build will include a manifest of which sub-modules the MFE provides at runtime.
+3. ``@openedx/frontend-build`` will specify - through its Webpack configs - a common set of shared dependencies to be used at runtime by all MFEs.
+4. Webpack will intelligently resolve those dependencies at runtime, `taking into account each module's specific version requirements <https://www.angulararchitects.io/en/blog/getting-out-of-version-mismatch-hell-with-module-federation>`_.
+5. MFEs can dynamically load modules from other MFEs at runtime with Webpack handling hooking them up to the right dependencies.
+
+Because we already use Webpack, the work to add the ModuleFederationPlugin to our configurations is small and uninvasive (see proof of concept in the `Reference Implementation`_ section below).
+
+Maintaining Dependency Consistency
+----------------------------------
+
+This proposal fundamentally changes how we work with MFE dependencies, and will require us to adopt a more rigorous approach to ensuring dependency consistency and compatiblity. That consistency doesn't come for free just by adopting shared dependencies at runtime: independent codebases will need to be kept in sync.
+
+We expect that this may need to take a number of possible forms.
```

Comment:
Language concern: this is an OEP, fundamental foundational guiding document. It should serve as a guide to How We Do Things. "We propose", "this should happen", "we need a process", etc doesn't feel like language for an OEP. I'd prefer stuff that's in flux point at a more mutatable document, such as a wiki page, where we can iterate and change quickly and eventually update the oep (or add an ADR) with This Is How We Do Things. But honestly, an OEP can very much just assert how process needs to change! Like a blueprint for the thing we're building, that isn't built yet. That may require making changes in other places (other OEPs, ADRs, wiki pages, documentation pages), and that's fine.

For example the Process section below could say:

Maintainers and developers require a way of knowing what dependency versions to use and when they need to upgrade to stay consistent. Thus, Open edX release documentation must include documentation of which frontend dependency versions are compatible with the release, likely pinned to a major version (i.e., React 17.x, Paragon 22.x, etc.).

Following the process defined in  the `Upgrade Project Runbook <https://openedx.atlassian.net/wiki/spaces/AC/pages/3660316693/Upgrade+Project+Runbook>`_, core repositories (as defined by OEP-57, Core Product) will be migrated through breaking changes in third-party dependencies.

--

That will require updates to the Upgrade Project Runbook, probably some BTR documentation, maybe other stuff. That's great. We should assert the new process in this doc, where - crucially - the community can debate and agree upon it, and then make it happen!